### PR TITLE
Closes #36 | Add COPAL dataloader

### LINKED ISSUE #36: Create dataset loader for COPAL
Dataloader name: `copal/copal.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?copal

|     Dataset| copal |
|-------------|---|
| Description | a novel Indonesian language common sense reasoning dataset. Unlike the previous Indonesian COPA dataset (XCOPA-ID), COPAL-ID incorporates Indonesian local and cultural nuances, and therefore, provides a more natural portrayal of day-to-day causal reasoning within the Indonesian cultural sphere. Professionally written by natives from scratch, COPAL-ID is more fluent and free from awkward phrases, unlike the translated XCOPA-ID. In addition, we present COPAL-ID in both standard Indonesian and in Jakartan Indonesian–a dialect commonly. It consist of premise, choice1, choice2, question, and label just like COPA dataset. |
| Subsets | test, test_colloquial |
| Languages | ind |
| Tasks | Commonsense Reasoning |
| License | Creative Commons Attribution 4.0 (cc-by-4.0) |
| Homepage | https://huggingface.co/datasets/haryoaw/COPAL |
| HF URL | https://huggingface.co/datasets/haryoaw/COPAL |
| Paper URL | https://arxiv.org/pdf/2311.01012v1.pdf |

### PR DESCRIPTION
**Pull request by williamnixon20**
Closes [#36](https://github.com/SEACrowd/seacrowd-datahub/issues/36)

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: seacrowd/sea_datasets/copal/copal.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/copal/copal.py:34]: btw, we're gonna introduce `COMMONSENSE_REASONING` task under this [PR](https://github.com/SEACrowd/seacrowd-datahub/pull/85). please wait for that PR and re-implement once it has been approved and merged. The changes will be minimum tho, you just need to add `meta` column in SEACrowd Schema

**Review by sabilmakbar — CHANGES_REQUESTED** [review]
The code looks good however I haven't tested it yet using test-suites! Just minor changes and few comments for now, and I'll update with test-suites result later

File: seacrowd/sea_datasets/copal/copal.py
Comment (williamnixon20) [reply seacrowd/sea_datasets/copal/copal.py:34]: got it, thanks!

Comment (sabilmakbar) [reply seacrowd/sea_datasets/copal/copal.py:34]: Hi @williamnixon20, the PR has been merged recently. Please pull it to your branch/fork and re-implement it. I can help on guiding that if needed 😃

Comment (williamnixon20) [reply seacrowd/sea_datasets/copal/copal.py]: Thanks for offer, appreciate it!  I already added the meta tag and changed the task. As you said, the changes were minimum 😄

File: seacrowd/sea_datasets/copal/copal.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/copal/copal.py:98]: since `COPAL` has columns like `terminology`, `culture`, and `language` which can be useful for evaluation, do you mind to add those into `features["meta"]` field?

it can be added as follows:
```python
features["meta"] = {
   'terminology': datasets.Value("int64"),
   'culture': datasets.Value("int64"),
   'language': datasets.Value("int64")}
```

and add the data aswell in `_generate_examples()` like:
```python
"meta": {
    "terminology": row.Terminology,
    "culture": row.Culture,
    "language": row.Language
}
```

File: seacrowd/sea_datasets/copal/copal.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/copal/copal.py:44]: slight req: can you slightly change the class name from `copal` to `COPAL`, since it will follows CamelCase rule of class naming (and COPAL itself is an acronym)

**Review by sabilmakbar — CHANGES_REQUESTED** [review]

File: seacrowd/sea_datasets/copal/copal.py
Comment (williamnixon20) [reply seacrowd/sea_datasets/copal/copal.py:98]: done, thanks for the snippets 👍

**Comment by williamnixon20** [comment]
btw i don't know why people were tagged for reviews. did I do something?

**Comment by sabilmakbar** [comment]
> btw i don't know why people were tagged for reviews. did I do something?

nah that's fine, perhaps due to the change in `.github/CODEOWNERS`

File: seacrowd/sea_datasets/copal/copal.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/copal/copal.py:16]: this one shd be okay if kept as lowercase, kinda weird to have this value as uppercase, even more this is being defined for URLS...

File: seacrowd/sea_datasets/copal/copal.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/copal/copal.py:30]: I think this is less intuitive, can we keep this as single-level dict? just remove the `_DATASETNAME`, wdyt?

File: seacrowd/sea_datasets/copal/copal.py
Comment (sabilmakbar) [inline seacrowd/sea_datasets/copal/copal.py:51]: this is a hardcoded one, can we make it less hardcoded by defining as `f"{_DATASETNAME}_source"` or `f"{_DATASETNAME}_seacrowd_qa"`?

**Review by sabilmakbar — CHANGES_REQUESTED** [review]
few more minor suggestions :)

File: seacrowd/sea_datasets/copal/copal.py
Comment (williamnixon20) [reply seacrowd/sea_datasets/copal/copal.py:30]: agree lol. done

**Comment by sabilmakbar** [comment]
Passed tests of the following:
1. Test Suite check
```bash
python -m tests.test_seacrowd seacrowd/sea_datasets/copal/copal.py
```
2. All configs check (executed in `base_dir` of SEACrowd Datahub):
```python
from datasets import load_dataset
from seacrowd.sea_datasets.copal.copal import COPAL

copal_class = COPAL()

parent_dir = "seacrowd/sea_datasets"
dset_name = "copal"
path_name = f"{parent_dir}/{dset_name}/{dset_name}.py"
for config_name in copal_class.builder_configs.keys():
    print(f"Using config name of: '{config_name}'")
    dset = load_dataset(path_name, name = config_name, download_mode="force_redownload")
```

**Review by sabilmakbar — APPROVED** [review]
LGTM! Thanks for implementing this dataloader! Let's wait for @gentaiscool's review

**Comment by SamuelCahyawijaya** [comment]
cc @gentaiscool

**Comment by sabilmakbar** [comment]
Hi @holylovenia @SamuelCahyawijaya , since @gentaiscool haven't got the time to check on this, any other ideas on how to proceed w/ this PR?

**Comment by holylovenia** [comment]
> Hi @holylovenia @SamuelCahyawijaya , since @gentaiscool haven't got the time to check on this, any other ideas on how to proceed w/ this PR?

Thanks for the heads up, @sabilmakbar. I'm assigning myself to replace @gentaiscool for this dataloader. Please wait for my review. 🙏

**Review by holylovenia — APPROVED** [review]
LGTM too!! Thanks a lot for your dataloader, @williamnixon20! 💪 Great work. I'm merging this.

**Pull request merged**
